cli_tools/gce_windows_upgrade/upgrader/upgrader.go

```diff
@@ -0,0 +1,528 @@
+	// 'windows-startup-script-url' exists, backup it
+	if windowsStartupScriptURLBackup != nil {
+		fmt.Printf("\nDetected current '%v', value='%v'. Will backup to '%v'.\n\n", metadataKeyWindowsStartupScriptURL,
+			*windowsStartupScriptURLBackup, metadataKeyWindowsStartupScriptURLBackup)
+		preparationVarMap["original_startup_script_url"] = *windowsStartupScriptURLBackup
+	}
+	prepWf, err := daisycommon.ParseWorkflow(upgradePreparationWorkflowPath, preparationVarMap,
+		params.project, params.zone, params.ScratchBucketGcsPath, params.Oauth, params.Timeout,
+		params.Ce, params.GcsLogsDisabled, params.CloudLogsDisabled, params.StdoutLogsDisabled)
+	if err != nil {
+		return nil, err
+	}
+	if windowsStartupScriptURLBackup == nil {
+		if !windowsStartupScriptURLBackupExists {
+			fmt.Printf("\nNo existing '%v' detected. Won't backup it.\n\n", metadataKeyWindowsStartupScriptURL)
+		}
+		// remove 'backup-script' step
+		delete(prepWf.Steps, "backup-script")
+		delete(prepWf.Dependencies, "backup-script")
+		prepWf.Dependencies["set-script"] = []string{"attach-install-disk"}
+	}
+
+	if params.SkipMachineImageBackup {
+		// remove 'backup-machine-image' step
+		delete(prepWf.Steps, "backup-machine-image")
+		delete(prepWf.Dependencies, "backup-machine-image")
+		prepWf.Dependencies["backup-os-disk-snapshot"] = []string{"stop-instance"}
+	}
+
+	err = daisyutils.RunWorkflowWithCancelSignal(ctx, prepWf)
+	return prepWf, err
+}
+
+func upgrade(ctx context.Context, workflowPath string, upgradeVarMap map[string]string,
+	params *UpgradeParams) (*daisy.Workflow, error) {
+
+	if windowsStartupScriptURLBackup != nil {
+		upgradeVarMap["original_startup_script_url"] = *windowsStartupScriptURLBackup
+	}
+
+	upgradeWf, err := daisycommon.ParseWorkflow(workflowPath, upgradeVarMap,
+		params.project, params.zone, params.ScratchBucketGcsPath, params.Oauth, params.Timeout,
+		params.Ce, params.GcsLogsDisabled, params.CloudLogsDisabled, params.StdoutLogsDisabled)
+	if err != nil {
+		return nil, err
+	}
+
+	err = daisyutils.RunWorkflowWithCancelSignal(ctx, upgradeWf)
+	return upgradeWf, err
+}
+
+func cleanup(ctx context.Context, cleanupVarMap map[string]string, params *UpgradeParams) (*daisy.Workflow, error) {
+
+	cleanupWf, err := daisycommon.ParseWorkflow(cleanupWorkflowPath, cleanupVarMap,
+		params.project, params.zone, params.ScratchBucketGcsPath, params.Oauth, params.Timeout,
+		params.Ce, params.GcsLogsDisabled, params.CloudLogsDisabled, params.StdoutLogsDisabled)
+	if err != nil {
+		return nil, err
+	}
+
+	err = daisyutils.RunWorkflowWithCancelSignal(ctx, cleanupWf)
+	return cleanupWf, err
+}
+
+func rollback(ctx context.Context, params *UpgradeParams, installMediaDiskName, newOSDiskName string) (*daisy.Workflow, error) {
+	originalStartupScriptURL := ""
+	if windowsStartupScriptURLBackup != nil {
+		originalStartupScriptURL = *windowsStartupScriptURLBackup
+	}
+
+	rollbackWf := &daisy.Workflow{
```

Comment:
Instead of creating it like this you can just use the NewStep methods, that's why they are there.

Reply:
done